backend/common/src/main/java/ai/verta/modeldb/common/db/migration/Migrator.java

```diff
@@ -22,19 +31,150 @@ public Migrator(Connection connection, String resourcesDirectory) {
+              Migration::isUp,
+              migration ->
+                  (migration.getNumber() > currentState.getVersion())
+                      && (migration.getNumber() <= versionToTarget));
+    } else {
+      migrationsToPerform =
+          gatherMigrations(
+                  Migration::isDown,
+                  migration ->
+                      migration.getNumber() > versionToTarget
+                          && migration.getNumber() <= currentState.getVersion())
+              .descendingSet();
+    }
+    return migrationsToPerform;
+  }
+
+  private MigrationDatastore setupDatastore(RdbConfig config) throws SQLException {
+    MigrationDatastore migrationDatastore = MigrationDatastore.create(config, connection);
     migrationDatastore.ensureMigrationTableExists();
+    MigrationState currentState = findCurrentState();
+    if (currentState == null) {
+      try (PreparedStatement ps =
+          connection.prepareStatement(
+              "INSERT INTO schema_migrations (version, dirty) values (?, ?) ")) {
+        ps.setInt(1, 0);
+        ps.setBoolean(2, false);
+        int rowsInserted = ps.executeUpdate();
+        if (rowsInserted != 1) {
+          throw new IllegalStateException(
+              "Failed to insert initial row into the schema_migrations table. rowsInserted: "
+                  + rowsInserted);
+        }
+      }
+    }
+    return migrationDatastore;
+  }
 
+  private void runMigrations(
+      MigrationDatastore migrationDatastore, NavigableSet<Migration> migrationsToPerform)
+      throws MigrationException, SQLException {
     MigrationTools.lockDatabase(migrationDatastore);
     try {
-      // todo: actually do some migration work
+      for (Migration migration : migrationsToPerform) {
+        try {
+          updateVersion(migration, true);
+          executeSingleMigration(migration);
+          updateVersion(migration, false);
+        } catch (IOException | SQLException e) {
+          throw new MigrationException("failed migration " + migration + "", e);
+        }
+      }
     } finally {
       migrationDatastore.unlock();
     }
   }
 
-  void executeMigration(Migration migration) throws IOException, SQLException {
+  private void updateVersion(Migration pendingMigration, boolean dirty) throws SQLException {
+    try (PreparedStatement ps =
+        connection.prepareStatement("UPDATE schema_migrations set version = ?, dirty = ?")) {
+      ps.setInt(
+          1,
+          pendingMigration.isUp()
+              ? pendingMigration.getNumber()
+              : pendingMigration.getNumber() - 1);
+      ps.setBoolean(2, dirty);
+      int rowsUpdated = ps.executeUpdate();
+      if (rowsUpdated != 1) {
+        throw new IllegalStateException(
+            "Failed to update schema_migrations table. rowsUpdated: " + rowsUpdated);
+      }
+    }
+  }
+
+  private NavigableSet<Migration> gatherMigrations(
+      Predicate<Migration> upOrDown, Predicate<Migration> versionPredicate)
```

Comment:
Suggest different names.
```suggestion
      Predicate<Migration> migrationDirection, Predicate<Migration> migrationMatchesVersionRange)
```

Reply:
filter rather than predictate?

Reply:
My thinking was that the names should indicate intended use/meaning.  It's less about `Filter` vs `Predicate` and more that the rest of the names didn't feel like they helped me much. 

Though now that I think about it probably makes just as much sense to have a `Predicate<Migration>... migrationFilters`.  That would simplify the call in `findVersionToTarget()`. :shrug: